src/bif/map.rs

```diff
@@ -44,33 +55,85 @@ pub fn bif_maps_is_key_2(_vm: &vm::Machine, process: &RcProcess, args: &[Value])
         let exist = hamt_map.find(target).is_some();
         return Ok(Value::boolean(exist))
     }
-    let heap = &process.context_mut().heap;
-    let tuple = tup2!(&heap, atom!(BADMAP), map.clone());
-    Err(Exception::with_value(Reason::EXC_BADMAP, tuple))
+    Err(Exception::with_value(Reason::EXC_BADMAP, map.clone()))
 }
 
-pub fn bif_maps_keys_1(_vm: &vm::Machine, _process: &RcProcess, args: &[Value]) -> BifResult {
-    unimplemented!();
+pub fn bif_maps_keys_1(_vm: &vm::Machine, process: &RcProcess, args: &[Value]) -> BifResult {
+    let map = &args[0];
+    if let Value::Map(m) = map {
+        let hamt_map = &m.0;
+        let heap = &process.context_mut().heap;
+        let mut vec = vec![];
+        for (key, _) in hamt_map.iter() {
+            vec.push(key.clone());
+        }
+        let list = vec_to_list!(heap, vec);
+        return Ok(list);
+    }
+    Err(Exception::with_value(Reason::EXC_BADMAP, map.clone()))
 }
 
 pub fn bif_maps_merge_2(_vm: &vm::Machine, _process: &RcProcess, args: &[Value]) -> BifResult {
     unimplemented!();
 }
 
-pub fn bif_maps_put_3(_vm: &vm::Machine, _process: &RcProcess, args: &[Value]) -> BifResult {
-    unimplemented!();
+pub fn bif_maps_put_3(_vm: &vm::Machine, process: &RcProcess, args: &[Value]) -> BifResult {
+    let map = &args[0];
+    let key = &args[1];
+    let value = &args[2];
+    if let Value::Map(m) = map {
+        let map = &(*m.0);
+        let new_map = map.clone().plus(key.clone(), value.clone());
```

Comment:
I have got 
   |                       cannot move out of `*map` which is behind a `&` reference
   |                       `map` is a `&` reference, so the data it refers to cannot be moved
by simply using map.plus()

Reply:
Ahh I see, https://docs.rs/hamt-rs/0.3.0/hamt_rs/struct.HamtMap.html#method.plus says that the method signature takes `self` by value instead of reference (`&self`). I think we can't get around that clone then